keps/sig-node/2008-forensic-container-checkpointing/README.md

```diff
@@ -0,0 +1,248 @@
+forensically analyse a copy of the container to understand if it might
+have been a possible threat. As the analysis is happening on a copy of
+the original container a possible attacker of the original container
+will not be aware of any sandboxed analysis.
+
+### Goals
+
+The goal of this KEP is to introduce *checkpoint* and *restore* to the CRI API.
+This includes extending the *kubelet* API to support checkpointing single
+containers with the User Story in mind (forensic use case).
+
+### Non-Goals
+
+Although *checkpoint* and *restore* can be used to implement container
+migration this KEP is only about enabling the forensic use case.
+
+## Proposal
+
+### Implementation
+
+For the forensic use case we want to offer the possibility to checkpoint a
+container out of a running Pod without stopping the checkpointed container and
+without letting the container know that it was checkpointed.
+
+The corresponding code changes for the forensic use case can be found in the
+following pull request:
+
+* https://github.com/kubernetes/kubernetes/pull/104907
+
+The goal is to introduce *checkpoint* and *restore* in a bottom-up approach.
+In a first step we only want to extend the CRI API to trigger a checkpoint
+by the container engine and to have the low level primitives in the *kubelet*
+to trigger a checkpoint. It is necessary to enable the feature gate
+`ContainerCheckpointRestore` to be able to checkpoint containers.
+
+In the corresponding pull request a checkpoint is triggered using the *kubelet*
+API:
+
+```
+curl -skv -X POST "https://localhost:10250/checkpoint/default/counters/wildfly"
+```
+
+For the first implementation we do not want to support restore in the
+*kubelet*. With the focus on the forensic use case the restore should happen
+outside of Kubernetes. The restore is a container engine only operation
+in this first step.
+
+The forensic use case is targeted to be part of the next (1.24) release.
+
+### User Stories
+
+To analyze unusual activities in a container, the container should
+be checkpointed without stopping the container or without the container
+knowing it was checkpointed. Using checkpointing it is possible to take
+a copy of a running container for forensic analysis. The container will
+continue to run without knowing a copy was created. This copy can then
+be restored in another (sandboxed) environment in the context of another
+container engine for detailed analysis of a possible attack.
+
+### Risks and Mitigations
+
+In its first implementation the risks are low as it tries to be a CRI API
+change with minimal changes to the kubelet and it is gated by the feature
+gate `ContainerCheckpointRestore`.
+
+## Design Details
+
+The feature gate `ContainerCheckpointRestore` will ensure that the API
+graduation can be done in the standard Kubernetes way.
+
+A kubelet API to trigger the checkpointing of a container will be
+introduced as described in [Implementation](#implementation).
+
```

Comment:
I have two comments: 
1. Different use cases would require the checkpoint to be done at different stages of the POD/container's life cycle or under some specific condition. To support all potential use cases, we should add an explicit "time" param in the checkpoint API. The "time" param can be 
	either a specific time in a container's life cycle, e.g. after create, before start, after start, before stop, after stop
	or when a condition is met, for example, container CPU utilization is less than 1% or there is no outbound traffic.

2. The checkpoint needs to be stored in someplace. It may be good to add an optional param in the API for specifying the checkpoint's storage location.

Reply:
I agree. That is correct, but we really want to introduce it as simple as possible to make it easy to review. Checkpoint/Restore is a complete new concept for Kubernetes and I think making the initial implementation as simple as possible, but still usable, is the best way to make it easy to review.

Adding additional parameters later is easy. Without getting the CRI API changes merged the work on the container engine level is difficult to test from Kubernetes. My main focus are really the CRI API changes.